### PR TITLE
add some fused kernels

### PR DESCRIPTION
**Pull request by dangkai4u**
Add some fused kernel used in Transformer:

- [x] fused_scale_mask_softmax_dropout: scale + mask + softmax + dropout
- [x] fused_scale_mask_softmax: scale + mask + softmax

### CODE REVIEW

File: oneflow/user/kernels/fused_scale_mask_softmax_dropout.cu [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    }
+  }
+  const SRC* src;
+  const int8_t* mask;
+  int64_t row_size;
+  SRC scale;
+};
+
+template<typename SRC, typename DST>
+struct DropoutScore {
```
Comment (MARD1NO) [inline oneflow/user/kernels/fused_scale_mask_softmax_dropout.cu:98]: Score -> Store

Comment (dangkai4u) [reply oneflow/user/kernels/fused_scale_mask_softmax_dropout.cu:98]: 已改，顺便问一下，functional_api.yaml中，对于xxx_grad函数也进行了导出，这个是不是没有必要，因为bind_python通常设置为False，而且用户大多数情况下也不会使用

Comment (hjchen2) [reply oneflow/user/kernels/fused_scale_mask_softmax_dropout.cu:98]: 在autograd里面会使用，具体看`oneflow/core/autograd/gradient_funcs/`

File: oneflow/core/autograd/gradient_funcs/fused_scale_mask_softmax.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+  ctx->SaveTensorForBackward(inputs.at(0));
+  return Maybe<void>::Ok();
+}
+
+Maybe<void> FusedScaleMaskSoftmax::Apply(const FusedScaleMaskSoftmaxInterState* ctx,
+                                         const TensorTuple& out_grads, TensorTuple* in_grads) const {
+  if (!ctx->input_requires_grad) { return Maybe<void>::Ok(); }
+
+  CHECK_EQ_OR_RETURN(out_grads.size(), 2);  // softmax_y, dy
```
Comment (MARD1NO) [inline oneflow/core/autograd/gradient_funcs/fused_scale_mask_softmax.cpp:65]: 这里的描述可能不太准确

fusedscalemaskSoftmax有两个输出，因此回来的梯度是两个

所以一个是softmax_y的梯度，一个是y的梯度。y的梯度叫dy可以，softmax_y的梯度这里注释需要重新写下，比如softmax_dy

File: oneflow/core/functional/impl/nn_functor.cpp [hunk 3]
```diff
@@ -1808,6 +1808,72 @@ class FusedBiasAddDropoutFunctor {
...
+    random_mask_like_op_ = CHECK_JUST(one::OpBuilder("random_mask_like").Input("like").Output("out").Build());
+    fused_scale_mask_softmax_dropout_op_ = CHECK_JUST(one::OpBuilder("fused_scale_mask_softmax_dropout")
+                                                          .Input("x")
+                                                          .Input("mask")
+                                                          .Input("dropout_mask")
+                                                          .Output("y")
+                                                          .Output("softmax_y")
+                                                          .Build());
+  }
+  Maybe<Tensor> operator()(const std::shared_ptr<one::Tensor>& x, const std::shared_ptr<one::Tensor>& mask,
```
Comment (MARD1NO) [inline oneflow/core/functional/impl/nn_functor.cpp:1844]: 你最后调用的op是 `fused_scale_mask_softmax_dropout_op_`，他有两个输出，所以这里返回值应该是一个Tensor Tuple。不然最后输出只有一个，你在求导那里保存不到softmax_y，进而求不了导数

File: oneflow/core/autograd/gradient_funcs/fused_scale_mask_softmax.cpp (hunk 2 above)
Comment (guo-ran) [reply oneflow/core/autograd/gradient_funcs/fused_scale_mask_softmax.cpp:65]: softmax_y 会输出给前向吗得到diff吗，印象中一般是用来传给后向op？

Comment (MARD1NO) [reply oneflow/core/autograd/gradient_funcs/fused_scale_mask_softmax.cpp:65]: 传过来应该都是两个梯度，但我们只要dy的梯度。softmax_y是传给后向op的，这里我只是提醒下这里的注释什么的

File: oneflow/core/functional/functional_api.yaml [hunk 4]
```diff
@@ -1487,12 +1487,28 @@
...
+  signature: "Tensor (Tensor softmax_y, Tensor dy, Tensor dropout_mask, Tensor mask, Float scale=1.0, Float dropout_scale=1.0) => FusedScaleMaskSoftmaxDropoutGrad"
+  bind_python: False
+
 - name: "fused_scale_tril_softmax_mask_scale"
   signature: "TensorTuple (Tensor a, *, Float p=0.5, Int64 diagonal, Float tril_scale_value, Generator generator=None) => FusedScaleTrilSoftmaxMaskScale"
   bind_python: True
 
 - name: "fused_scale_tril_softmax_mask_scale_grad"
-  signature: "Tensor (Tensor softmax_y, Tensor dy, Tensor mask, Int64 diagonal, Float tril_scale_value, Float mask_scale_value) => FusedScaleTrilSoftmaxMaskScaleGrad"
+  signature: "Tensor (Tensor y, Tensor dy, Tensor mask, Int64 diagonal, Float tril_scale_value, Float mask_scale_value) => FusedScaleTrilSoftmaxMaskScaleGrad"
```
Comment (MARD1NO) [inline oneflow/core/functional/functional_api.yaml:1511]: 这个不用改吧，用的就是softmax_y？

File: oneflow/core/autograd/gradient_funcs/fused_scale_mask_softmax_dropout.cpp [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+                                                  const AttrMap& attrs) const {
+  CHECK_EQ_OR_RETURN(inputs.size(), 3); // input, mask, dropout_mask
+  ctx->input_requires_grad = inputs.at(0)->requires_grad();
+
+  if (!ctx->input_requires_grad) { return Maybe<void>::Ok(); }
+  ComposedAttrMap composed_attrs(attrs, base_attrs_);
+  ctx->scale = JUST(composed_attrs.GetAttr<float>("scale_value"));
+  ctx->dropout_scale = JUST(composed_attrs.GetAttr<float>("dropout_scale_value"));
+
+  ctx->SaveTensorForBackward(inputs.at(1));
```
Comment (MARD1NO) [inline oneflow/core/autograd/gradient_funcs/fused_scale_mask_softmax_dropout.cpp:63]: 可以加点注释，表示这里save了什么

File: oneflow/user/kernels/fused_scale_mask_softmax.cu [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+  }
+  const SRC* src;
+  const int8_t* mask;
+  int64_t row_size;
+  SRC fill;
+  SRC scale;
+};
+
+template<typename SRC, typename DST>
+struct ScaleMaskStore {
```
Comment (MARD1NO) [inline oneflow/user/kernels/fused_scale_mask_softmax.cu:49]: load是softmax之前的操作，store是softmax之后的操作，如果你没有dropout后，那么store就应该什么操作都不需要吧

File: oneflow/core/functional/functional_api.yaml (hunk 4 above)
Comment (dangkai4u) [reply oneflow/core/functional/functional_api.yaml]: 好的，这里改了

**Comment by dangkai4u** [comment]
这个pr我本地测试通过了，你们可以测试一下，之后就合并吧

File: oneflow/user/ops/fused_scale_mask_softmax_dropout_op.cpp [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+      const user_op::TensorDesc& x_desc = ctx->InputTensorDesc("x", 0);
+      *ctx->OutputShape("y", 0) = x_desc.shape();
+      *ctx->OutputIsDynamic("y", 0) = x_desc.is_dynamic();
+      *ctx->OutputShape("softmax_y", 0) = x_desc.shape();
+      *ctx->OutputIsDynamic("softmax_y", 0) = x_desc.is_dynamic();
+      return Maybe<void>::Ok();
+    })
+    .SetDataTypeInferFn([](user_op::InferContext* ctx) -> Maybe<void> {
+      const user_op::TensorDesc& x_desc = ctx->InputTensorDesc("x", 0);
+      *ctx->OutputDType("y", 0) = x_desc.data_type();
```
Comment (MARD1NO) [inline oneflow/user/ops/fused_scale_mask_softmax_dropout_op.cpp:41]: 需要加个check，保证mask的dtype是int8类型

File: oneflow/user/ops/fused_scale_mask_softmax_dropout_op.cpp [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    .Input("x")
+    .Input("mask")
+    .Input("dropout_mask")
+    .Output("y")
+    .Output("softmax_y")
+    .Attr<float>("scale_value", 1.0)
+    .Attr<float>("mask_fill_value", 0.)
+    .Attr<float>("dropout_scale_value", 1.0)
+    .SetTensorDescInferFn([](user_op::InferContext* ctx) -> Maybe<void> {
+      const user_op::TensorDesc& x_desc = ctx->InputTensorDesc("x", 0);
```
Comment (MARD1NO) [inline oneflow/user/ops/fused_scale_mask_softmax_dropout_op.cpp:32]: 现在mask和x是elementwise关系，所以要check mask的shape是不是和x的shape一致

File: oneflow/user/ops/fused_scale_mask_softmax_dropout_op.cpp [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+            .Split(user_op::OpArg("softmax_y", 0), axis)
+            .Build();
+      }
+      return Maybe<void>::Ok();
+    });
+
+REGISTER_USER_OP("fused_scale_mask_softmax_dropout_grad")
+    .Input("softmax_y")
+    .Input("dy")
+    .Input("dropout_mask")
```
Comment (MARD1NO) [inline oneflow/user/ops/fused_scale_mask_softmax_dropout_op.cpp:73]: 前面顺序是先mask，然后dropout_mask，这里也保持顺序一致吧，然后functional那里也要记得改

File: oneflow/user/ops/fused_scale_mask_softmax_dropout_op.cpp [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    .Input("mask")
+    .Output("dx")
+    .Attr<float>("scale_value")
+    .Attr<float>("dropout_scale_value")
+    .SetTensorDescInferFn([](user_op::InferContext* ctx) -> Maybe<void> {
+      const user_op::TensorDesc& softmax_y_desc = ctx->InputTensorDesc("softmax_y", 0);
```
Comment (MARD1NO) [inline oneflow/user/ops/fused_scale_mask_softmax_dropout_op.cpp:79]: 同上加入关于mask的check

File: oneflow/user/ops/fused_scale_mask_softmax_dropout_op.cpp [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+                .Build();
+        op.BindGradTensorWithOpInput(grad_op.output("dx", 0), "x", 0);
+        AddOp(grad_op);
+      }
+      return Maybe<void>::Ok();
+    });
+
+}  // namespace
+
+}  // namespace oneflow
```
Comment (MARD1NO) [inline oneflow/user/ops/fused_scale_mask_softmax_dropout_op.cpp:133]: 换行，后面的文件也要这么换行

File: oneflow/user/ops/fused_scale_mask_softmax_op.cpp [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+namespace {
+
+REGISTER_USER_OP("fused_scale_mask_softmax")
+    .Input("x")
+    .Input("mask")
+    .Output("y")
+    .Attr<float>("scale_value", 1.0)
+    .Attr<float>("mask_fill_value", 0.)
+    .SetTensorDescInferFn([](user_op::InferContext* ctx) -> Maybe<void> {
+      const user_op::TensorDesc& x_desc = ctx->InputTensorDesc("x", 0);
```
Comment (MARD1NO) [inline oneflow/user/ops/fused_scale_mask_softmax_op.cpp:29]: 相关check也要补充

File: python/oneflow/test/modules/test_fused_scale_mask_softmax_dropout.py [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+def _test_fused_scale_mask_softmax_dropout(
+    test_case, batch_size, num_heads, seq_length, seed, fill_value, scale_value, p,
+):
+    generator = flow.Generator()
+    generator.manual_seed(seed)
+
+    x = np.random.randn(batch_size, num_heads, seq_length, seq_length)
+    mask = np.random.randint(0, 2, size=(batch_size, num_heads, seq_length, seq_length), dtype=np.uint8)
+
+    # x = np.array([[1., 2., 3.], [4., 5., 6.]])
```
Comment (MARD1NO) [inline python/oneflow/test/modules/test_fused_scale_mask_softmax_dropout.py:34]: 注释去除

File: oneflow/core/autograd/gradient_funcs/fused_scale_mask_softmax.cpp [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+                                           const TensorTuple& inputs, const TensorTuple& outputs,
+                                           const AttrMap& attrs) const {
+  CHECK_EQ_OR_RETURN(inputs.size(), 2); // input, mask
+  ctx->input_requires_grad = inputs.at(0)->requires_grad();
+
+  if (!ctx->input_requires_grad) { return Maybe<void>::Ok(); }
+  ComposedAttrMap composed_attrs(attrs, base_attrs_);
+  ctx->scale = JUST(composed_attrs.GetAttr<float>("scale_value"));
+
+  ctx->SaveTensorForBackward(inputs.at(1));
```
Comment (MARD1NO) [inline oneflow/core/autograd/gradient_funcs/fused_scale_mask_softmax.cpp:61]:
```
ctx->SaveTensorForBackward(inputs.at(1)); / / Save mask
```
这样补充注释

File: oneflow/core/autograd/gradient_funcs/fused_scale_mask_softmax.cpp [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+  ctx->SaveTensorForBackward(outputs.at(0)); // save y, ie. softmax result
```
Comment (MARD1NO) [inline oneflow/core/autograd/gradient_funcs/fused_scale_mask_softmax.cpp:62]: 同上

File: oneflow/core/autograd/gradient_funcs/fused_scale_mask_softmax_dropout.cpp (hunk 5 above)
Comment (MARD1NO) [reply oneflow/core/autograd/gradient_funcs/fused_scale_mask_softmax_dropout.cpp:63]: 这里也要加

File: oneflow/core/functional/impl/nn_grad_functor.cpp [hunk 16]
```diff
@@ -806,6 +806,50 @@ class FusedScaleTrilSoftmaxMaskScaleGradFunctor {
...
+  std::shared_ptr<OpExpr> op_;
+};
+
+class FusedScaleMaskSoftmaxDropoutGradFunctor {
+ public:
+  FusedScaleMaskSoftmaxDropoutGradFunctor() {
+    op_ = CHECK_JUST(one::OpBuilder("fused_scale_mask_softmax_dropout_grad")
+                         .Input("softmax_y")
+                         .Input("dy")
+                         .Input("dropout_mask")
```
Comment (MARD1NO) [inline oneflow/core/functional/impl/nn_grad_functor.cpp:835]: 顺序还是按照先mask后dropmask，跟前向顺序保持一致

File: oneflow/core/functional/functional_api.yaml [hunk 17]
```diff
@@ -1487,6 +1487,22 @@
   signature: "Tensor (Tensor a, Tensor b, *, Float p=0.5, Int32 axis, Generator generator=None) => FusedBiasAddDropout"
   bind_python: True
 
+- name: "fused_scale_mask_softmax"
+  signature: "Tensor (Tensor x, Tensor mask, *, Float fill_value, Float scale=1.0) => FusedScaleMaskSoftmax"
```
Comment (MARD1NO) [inline oneflow/core/functional/functional_api.yaml:1491]: fill_value要不要给个默认值，我觉得0.0是不是比较合理

File: python/oneflow/test/modules/test_fused_scale_mask_softmax.py [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+        np.allclose(fused_out.numpy(), origin_out.numpy(), atol=1e-4, rtol=1e-4)
+    )
+    test_case.assertTrue(
+        np.allclose(fused_x_tensor.grad.numpy(), origin_x_tensor.grad.numpy(), atol=1e-4, rtol=1e-4)
+    )
+
+@flow.unittest.skip_unless_1n1d()
+@unittest.skipIf(os.getenv("ONEFLOW_TEST_CPU_ONLY"), "only test gpu cases")
+class TestFusedScaleMaskSoftmax(flow.unittest.TestCase):
+    def test_gather(test_case):
```
Comment (MARD1NO) [inline python/oneflow/test/modules/test_fused_scale_mask_softmax.py:59]: 这个def的名字改一下

File: python/oneflow/test/modules/test_fused_scale_mask_softmax.py [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+import oneflow as flow
+import oneflow.unittest
+
+
+def _test_fused_scale_mask_softmax(
+    test_case, batch_size, num_heads, seq_length, seed, fill_value, scale_value,
+):
+
+    x = np.random.randn(batch_size, num_heads, seq_length, seq_length)
+    mask = np.random.randint(1, 2, size=(batch_size, num_heads, seq_length, seq_length), dtype=np.uint8)
```
Comment (MARD1NO) [inline python/oneflow/test/modules/test_fused_scale_mask_softmax.py:30]: 是不是应该是randint(0, 1)才对

File: python/oneflow/test/modules/test_fused_scale_mask_softmax_dropout.py [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+        np.allclose(fused_out.numpy(), origin_out.numpy(), atol=1e-4, rtol=1e-4)
+    )
+    test_case.assertTrue(
+        np.allclose(fused_x_tensor.grad.numpy(), origin_x_tensor.grad.numpy(), atol=1e-4, rtol=1e-4)
+    )
+
+@flow.unittest.skip_unless_1n1d()
+@unittest.skipIf(os.getenv("ONEFLOW_TEST_CPU_ONLY"), "only test gpu cases")
+class TestFusedScaleMaskSoftmaxDropout(flow.unittest.TestCase):
+    def test_gather(test_case):
```
Comment (MARD1NO) [inline python/oneflow/test/modules/test_fused_scale_mask_softmax_dropout.py:64]: def名字换一下

**Review by MARD1NO — APPROVED** [review]

**Review by guo-ran — APPROVED** [review]
